usr/src/uts/armv8/io/arm_gtmr.c

```diff
@@ -0,0 +1,591 @@
+	 * contain an array of four integers and each integer must have the
+	 * value 14 (corresponding to CBE_HIGH_PIL), or the cyclic backend
+	 * will not behave correctly.
+	 */
+
+	if (!ddi_prop_exists(DDI_DEV_T_ANY, dip,
+	    DDI_PROP_DONTPASS, OBP_INTERRUPT_PRIORITIES)) {
+		if ((ret = ndi_prop_update_int_array(DDI_DEV_T_NONE, dip,
+		    OBP_INTERRUPT_PRIORITIES,
+		    def_prios, def_nprios)) != DDI_PROP_SUCCESS)
+			dev_err(dip, CE_PANIC,
+			    "Failed to set default priorities");
+	}
+
+	iwanted = 1;
+
+	inum = ARM_GTMR_INUM_PHYS_NS;
+	if (CPU->cpu_m.mcpu_boot_el == 1)
+		inum = ARM_GTMR_INUM_VIRTUAL;
+
+	/*
+	 * We can only use fixed interrupts
+	 */
+	ret = ddi_intr_get_supported_types(dip, &itypes);
+	if ((ret != DDI_SUCCESS) || (!(itypes & DDI_INTR_TYPE_FIXED)))
+		dev_err(dip, CE_PANIC, "Fixed type interrupt is not supported");
+
+	/*
+	 * We expect four fixed interrupts, but accept any number that allows
+	 * us to use our expected interrupt number
+	 */
+	ret = ddi_intr_get_nintrs(dip, DDI_INTR_TYPE_FIXED, &icount);
+	if (ret != DDI_SUCCESS)
+		dev_err(dip, CE_PANIC, "Failed to get fixed interrupt count");
+	if (icount < (inum + 1))
+		dev_err(dip, CE_PANIC, "Need %d fixed interrupts, found %d",
+		    inum + 1, icount);
+	if (icount != EXPECTED_NINTRS)
+		dev_err(dip, CE_NOTE, "Expected %d interrupts, found %d",
+		    EXPECTED_NINTRS, icount);
+
+	/*
+	 * Allocate our interrupt
+	 */
+	itable = kmem_zalloc(icount * sizeof (ddi_intr_handle_t), KM_SLEEP);
+	ret = ddi_intr_alloc(dip, itable, DDI_INTR_TYPE_FIXED,
+	    inum, iwanted, &iactual, DDI_INTR_ALLOC_STRICT);
+	if ((ret != DDI_SUCCESS) || (iactual != iwanted))
+		dev_err(dip, CE_PANIC, "ddi_intr_alloc() for %d fixed "
+		    "interrupts failed (inum %u, ret %d, actual %d)",
+		    iwanted, inum, ret, iactual);
+
+	/*
+	 * Ensure that our priority is correctly configured
+	 */
+	if (ddi_intr_get_pri(itable[inum], &intr_pri) != DDI_SUCCESS)
+		dev_err(dip, CE_PANIC, "ddi_intr_get_pri() failed");
+	VERIFY3U(intr_pri, ==, CBE_HIGH_PIL);
+
+	/*
+	 * Add our hardware interrupt handler to our interrupt handle.
+	 */
+	if ((ret = ddi_intr_add_handler(itable[inum], cbe_fire,
+	    NULL, NULL)) != DDI_SUCCESS)
+		dev_err(dip, CE_PANIC, "ddi_intr_add_handler() failed");
+
+	/*
+	 * Enable our primary per-processor interrupt. This enables the
+	 * interrupt for all processors, but the source remains masked, so
+	 * no interrupt will make it to a CPU just yet (see `cbe_enable').
+	 */
+	if ((ret = ddi_intr_enable(itable[inum])) != DDI_SUCCESS)
+		dev_err(dip, CE_PANIC, "ddi_intr_enable() failed");
+
+	/*
+	 * Add our xcall handler, also firing at priority 14. This is called
+	 * from our primary per-processor interrupt. In Arm terms this is
+	 * an SGI (software-generated interrupt) and, as such, has an
+	 * implementation-defined interrupt number.
+	 */
+	if (add_avintr(NULL, CBE_HIGH_PIL, cbe_fire_ipi, "cbe_fire_ipi",
+	    IRQ_IPI_CBE, 0, NULL, NULL, NULL) != 1)
```

Comment:
40a8a19 (interrupts: provide system PIC registration and dispatch) changed the last argument from `NULL` to `dip` so that the xcall IRQ is properly attributed in `::interrupts` - it might be worth folding that in here.

Reply:
Aah, I see you've added the second commit here - thanks!